x/payment/types/wirepayformessage.go

```diff
@@ -106,6 +109,10 @@ func (msg *MsgWirePayForMessage) ValidateBasic() error {
 			return err
 		}
 
+		if !powerOf2(commit.K) {
+			return fmt.Errorf("invalid square size, the size must be power of 2: %d", commit.K)
+		}
```

Comment:
can we move this to top of this for loop? I think it makes a lot more sense when reading the code. This check is also very quick, where creating the commitment is not, so it's good practice to check the quick things first before moving on to check the heavier things.

also, can we also add another [test case](https://github.com/toanngosy/celestia-app/blob/5498a551089919065e4165797bbb937768de9def/x/payment/types/payformessage_test.go#L291) for validate basic to make sure that this works? 

thanks

Reply:
depending on the size of the message, the commit may or may not change. The commit will only be different if the message is large enough to span more than a single row.

Reply:
I'm trying to understand [the test](https://github.com/toanngosy/celestia-app/blob/square-sizes/x/payment/types/payformessage_test.go#L297).

The tested message size is 1000. Each chunk will have a `ShareSize` of 256, so basically we need 4 shares to store the message (256, 256, 256, 232).

In the constructed wpfm, the first `SquareSize` is 16, so we have 16 rows and 16 cols. The message is not large enough to span more than a single row. In the test, we change the `SquareSize` to 128, the message is still less than a single row -> the commit does not change.

In the [message block layout](https://celestiaorg.github.io/celestia-specs/latest/rationale/message_block_layout.html), we also have "Messages that span multiple rows must begin at the start of a row".
Now, instead of 128, I change the `SquareSize` to 4. I think the `Q0` would be like this:
 _________________
|   X   |        |        |        |
|____|____|____|____|
| 256 | 256 | 256 | 232 |
|____|____|____|____|
|        |        |        |        |
|____|____|____|____|

Where `X` is the transaction paying for the message. 
I think it should have a different commit. However, the test [the test](https://github.com/toanngosy/celestia-app/blob/square-sizes/x/payment/types/payformessage_test.go#L297) does not throw back an error, which means the commit does not change.

Reply:
> I think this should have a different commit because the message is moved to a new row.

Why would the message be moved to a new row here? Wouldn't it always start on a new row? Then if it's only 1000, it won't actually span across more than one row.

> Where X is the transaction paying for the message.

The transactions shouldn't matter in this case, as the messages will always start on a new row.

If the message size is doubled to 2000, and the square size 16 is replaced with 4, the test passes. This is because the message is large enough to overflow onto the next row.

That being said, there are still some things wrong with the current state of the logic being used compared with what is supposed to occur as described in the specs, as we haven't actually implemented non interactive defaults yet [#455](https://github.com/celestiaorg/celestia-core/issues/455). Not only that, but the current logic also doesn't take into account the exact encoding logic that is used in celestia-core. We're currently working on a rather large refactor that will move much of this logic to the app, so hopefully after that, we will be able to write the code in a easier to grok way.

Reply:
Thanks, @evan-forbes, I changed the size to 2000 and square size 4. This PR should be good to go.
I will dig deeper to understand this, your comment above gives me lots of hints :D

Reply:
Awesome! Yeah, I would focus on what the specs say, as the implementation will eventually match those.